server/src/main/java/io/druid/curator/discovery/CuratorDruidLeaderSelector.java

```diff
@@ -0,0 +1,204 @@
+    this.curator = curator;
+    this.self = self;
+    this.latchPath = latchPath;
+  }
+
+  private LeaderLatch createNewLeaderLatch()
+  {
+    final LeaderLatch newLeaderLatch = new LeaderLatch(
+        curator, latchPath, self.getHostAndPortToUse()
+    );
+
+    newLeaderLatch.addListener(
+        new LeaderLatchListener()
+        {
+          @Override
+          public void isLeader()
+          {
+            try {
+              if (leader) {
+                log.warn("I'm being asked to become leader. But I am already the leader. Ignored event.");
+                return;
+              }
+
+              leader = true;
+              term++;
+              listener.becomeLeader();
+            }
+            catch (Exception ex) {
+              log.makeAlert(ex, "listener becomeLeader() failed. Unable to become leader").emit();
+
+              // give others a chance to become leader.
+              final LeaderLatch oldLatch = createNewLeaderLatch();
+              CloseQuietly.close(oldLatch);
+              leader = false;
+              try {
+                //Small delay before starting the latch so that others waiting are chosen to become leader.
+                Thread.sleep(1000);
+
+                leaderLatch.get().start();
+              }
+              catch (Exception e) {
+                // If an exception gets thrown out here, then the node will zombie out 'cause it won't be looking for
+                // the latch anymore.  I don't believe it's actually possible for an Exception to throw out here, but
+                // Curator likes to have "throws Exception" on methods so it might happen...
+                log.makeAlert(e, "I am a zombie").emit();
+              }
+            }
+          }
+
+          @Override
+          public void notLeader()
+          {
+            try {
+              if (!leader) {
+                log.warn("I'm being asked to stop being leader. But I am not the leader. Ignored event.");
+                return;
+              }
+
+              leader = false;
+              listener.stopBeingLeader();
+            }
+            catch (Exception ex) {
+              log.makeAlert(ex, "listener.stopBeingLeader() failed. Unable to stopBeingLeader").emit();
+            }
+          }
+        },
+        Execs.singleThreaded(StringUtils.format("LeaderSelector[%s]", latchPath))
+    );
+
+    return leaderLatch.getAndSet(newLeaderLatch);
+  }
+
+  @Override
+  public String getCurrentLeader()
+  {
+    if (!lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS)) {
+      throw new ISE("not started");
+    }
+
+    try {
+      final LeaderLatch latch = leaderLatch.get();
+
+      Participant participant = latch.getLeader();
+      if (participant.isLeader()) {
+        return participant.getId();
+      }
+
+      return null;
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public boolean isLeader()
+  {
+    return leader;
```

Comment:
if the leader status is being contested, should this block until the contest is settled?